### ISSUE TITLE
ModelicaTest.Fluid.TestComponents.Vessels.TestInitialization does not initialize

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/Modelica#1222]
**Reported by otter on 26 Jul 2013 13:29 UTC**
The model ModelicaTest.Fluid.TestComponents.Vessels.TestInitialization simulates with MSL 3.2 but initialization fails in MSL 3.2.1 when simulating with Dymola 2014 FD01 development, but initialization is successful with Dymola 2014.

With MSL 3.2, two linear systems of order 1 are present during initialization.

With MSL 3.2.1, additionally two **nonlinear** systems of order 2 are present during initialization (for both Dymola versions).

The issue seems to be related to the vessel equations with curve parameter s, because it appears in the nonlinear equations. Furthermore, the homotopy operator was newly introduced here.

It seems strange that nonlinear systems of equations occur in MSL 3.2.1 during initialization that had not been present previously and going back to the previous option with use_eps_Re = false does not change this (although use_eps_Re has an Evaluate=true annotation).

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/1222

**Comment by modelica-trac-importer** [comment]
**Comment by rfranke on 26 Jul 2013 17:37 UTC**
It would be surprising if the problem was related to `system.use_eps_Re` because its default setting of false keeps the equations from MSL 3.2 and because it shall be evaluated during model translation.

The mentioned equation systems disappear if the use of `homotopy` is removed from `Modelica.Fluid.Vessels.BaseClasses.PartialLumpedVessel` and from `Modelica.Fluid.Pipes.BaseClasses.FlowModels.PartialGenericPipeFlow`.

An obvious solution would be to somehow disable homotopy per default for backwards compatibility with MSL 3.2. Maybe people would like to have the option of not using homotopy also in a non-backwards compatible new version of MSL?

**Comment by modelica-trac-importer** [comment]
**Comment by fcasella on 26 Jul 2013 18:54 UTC**
Whether one wants to use homotopy or not can be easily decided by changing the appropriate flags in the Modelica tool (e.g. Advanced.SupportHomotopyFunction = false in Dymola)

BTW, Dymola 2014 by default first tries to solve the system without homotopy, and only if that fails homotopy is introduced (unless one sets Advanced.OnlyUseHomotopyMethod = true). I don't have access to Dymola 2014 FD01, so I don't know what happens there; maybe the default has changed to Advanced.OnlyUseHomotopyMethod = true?

As a final remark, it is strange that the simplified equations used in the homotopy operator actually result in more nonlinear equations to be solved. This should be investigated properly, but I'll only have time to deal with this starting in September.

**Comment by modelica-trac-importer** [comment]
**Comment by rfranke on 28 Jul 2013 05:34 UTC**
The equation systems are the same in Dymola 2014. Dymola 2014 solves them straightforward, without switching to homotopy mode. The point is that homotopy introduces new nonlinear equation systems, which hadn't been detected unless another version of Dymola now fails to solve them.

It appears that homotopy is still too little tested for an unconditional introduction in a backwards compatible MSL 3.2.1.

SVN fd0479c9e890aa6fd99c86358df8d5b996aeafb5 introduces the new function `Modelica.Fluid.Utilities.homotopic`. It is motivated by the trivial implementation given in the Modelica spec and makes the use of the homotopy operator configurable:
```
function homotopic "Conditional use of homotopy"
  extends Modelica.Icons.Function;
  input Real actual;
  input Real simplified;
  input Boolean use_homotopy;
  output Real y;
algorithm 
  y := if use_homotopy then homotopy(actual, simplified) else actual;
  annotation(Inline = true);
end homotopic;
```

All calls to homotopy have been replaced with calls to Utilities.homotopic in Modelica.Fluid, passing the additional new parameter system.use_homotopy:

![Image](System_use_homotopy.png)

The default value of false keeps the model equations as in MSL 3.2. Running all examples and tests without homotopy, there was one problem with `ModelicaTest.TestPipesAndValves.LumpedPipeInitialization`. I adjusted the valve configuration of this model to the configuration used by DynamicPipeInitialization. Both models should produce the same result, up to numerical issues -- configuring DynamicPipeInitialization.pipe to nNodes=1, modelStructure=a_v_b, the results should be identical.

The new `system.use_homotopy` provides a smooth introduction of this new feature, independent from advanced settings of specific versions of particular tools. People can start using homotopy. Remaining issues, like unexpected new equation systems, can be investigated after the release of MSL 3.2.1.

**Comment by modelica-trac-importer** [comment]
**Comment by rfranke on 29 Jul 2013 06:23 UTC**
SVN 30d46d73c2ef548a81fa3ab5b8d685e2a851dfdc adjusts the parameters and updates the docu of TestInitialization.

The model uses port outlets with a diameter of 0.25m. The volume has been increased from 0.1 to 100m3, so that the port outlets can be considered small, compared to the surface of the volume. 100m3 might be a living room.

The pipe lengths have been increased from 1 to 10m -- the pressure drops are so small that a shorter pipe should better be neglected. Also, 10m might be needed for a room with a volume of 100m3.

The model demonstrates the usefulness of system.use_eps_Re. Unfortunately the solver accuracy needed to be increased from 1e-4 (default) to 1e-6. This means that the automatic adjustment of nominal values can possibly still be improved.

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 30 Jul 2013 09:34 UTC**
Final fix in ee0583e28505e8721cad855f3e452bd4627edf5e:

- Undoing the fix of fd0479c9e890aa6fd99c86358df8d5b996aeafb5.

- According to a suggestion from Hans, slightly rewrote PartialLumpedVessel so that Boolean variables are used in the critical if-clause instead of Boolean expressions helping a tool to solve the mixed system of equations. So new formulation:
```
// fluid flow through ports
regularFlow[i] = fluidLevel >= portsData_height[i];
inFlow[i]      = not regularFlow[i] and 
                 (s[i] > 0 or portsData_height[i] >= fluidLevel_max);
if regularFlow[i] then
  // regular operation: fluidLevel is above ports[i]
  ...
elseif inFlow[i] then
  // ports[i] is above fluidLevel and has inflow
  ...
else
  // ports[i] is above fluidLevel, preventing outflow
  ...
end if;
```

With these changes, together with comment4 (30d46d73c2ef548a81fa3ab5b8d685e2a851dfdc; adjusting the parameters of the test model), simulation is now successful in Dymola (without applying the homotopy method, since initialization already solved by Newton solver).

The issue with the 2 nonlinear equations seems to be not an issue, but only an unclear statistics output of Dymola: When a homotopy operator is present, Dymola generates special code for the "simple case". For this "simple case" code the nonlinear (mixed) system of equation of the dynamic part is still present, since the homotopy simple case is not simple enough. This equation system is reported in the "initialization" section in the statistics.

**Comment by modelica-trac-importer** [comment]
**Modified by otter on 31 Jul 2013 15:51 UTC**
